### PR TITLE
Mget redis command

### PR DESCRIPTION
**Pull request by Valkyrie00**

### CODE REVIEW

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    if (mget_command_context->has_error) {
+        return true;
+    }
+
+    // The scenario tested in this assert can't happen, but if it does happen it's a bug in the protocol parser and
+    // there is no way to be sure that there is no corruption or data loss, so it's better to dramatically abort
+    // (in debug mode)
+    assert(mget_command_context->key_offset + chunk_length <= mget_command_context->key_length);
+
+    mget_command_context->key = mget_command_context->keys[argument_index] =
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: This allocation needs to be done in STREAM_BEGIN because STREAM_DATA can be called multiple times, especially with long keys the OS network stack of the client or the server can split up the data in multiple packets.

STREAM_DATA should only copy data data from the buffer of the chunk of data received into, in this case, the buffer of the key (line 123).

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        mget_command_context->has_error = true;
+        snprintf(
+                mget_command_context->error_message,
+                sizeof(mget_command_context->error_message) - 1,
+                "ERR The key has exceeded the allowed size of <%u>",
+                channel->protocol_config->redis->max_key_length);
+        return true;
+    }
+
+    mget_command_context->key = NULL;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Allocate the memory for the key here with `slab_allocator_mem_alloc(mget_command_context->key_length)`

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+NETWORK_PROTOCOL_REDIS_COMMAND_FUNCPTR_ARGUMENT_STREAM_END(mget) {
+    storage_db_entry_index_status_t old_status;
+    mget_command_context_t *mget_command_context = (mget_command_context_t*)protocol_context->command_context;
+
+    if (mget_command_context->has_error) {
+        goto end;
+    }
+
+    storage_db_entry_index_t *entry_index = storage_db_get_entry_index(
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: The `storage_db_entry_index_status_increase_readers_counter` call will force the garbage collector to carry out a later collection of the data if the entry gets deleted in the meantime, this wasn't a problem for the `get` command but for the `mget` it might be.

In the case an user request 100000 gets altogether (for whatever reason) the garbage collector will get stuck meanwhile the platform receives the keys that, with a lot of them, might become very time consuming.

The biggest problem though is that the entry_index is guaranteed to be available only for a very short arbitrary timeframe if `storage_db_entry_index_status_increase_readers_counter` is not invoked (to be more specific is guaranteed to be available until an I/O operation is not carried out, op which happens automatically at the end of the callback or when data are read/written from the storage or sent over the network)

The best would be to move this entire block of code in the loop in COMMAND END

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+    mget_command_context_t *mget_command_context = (mget_command_context_t*)protocol_context->command_context;
+    mget_command_context->has_error = false;
+    mget_command_context->keys = NULL;
+    mget_command_context->entries = NULL;
+    mget_command_context->entry_counter = 0;
+
+    // TODO
+    mget_command_context->keys = slab_allocator_mem_alloc((reader_context->arguments.count - 1) * sizeof(char*));
+    mget_command_context->entries = slab_allocator_mem_alloc((reader_context->arguments.count - 1) * sizeof(storage_db_entry_index_t*));
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: As per comments in STREAM_END, would be better to fetch and release the single entries on the spot (to reduce the pressure on the garbage collector) in COMMAND_END therefore it's not necessary to have an array of them.

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+struct mget_command_context {
+    char error_message[200];
+    bool has_error;
+    char *key;
+    size_t key_length;
+    size_t key_offset;
+    char **keys;
+    int entry_counter;
+    storage_db_entry_index_t **entries;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: As per comments below, the field `entries` can be dropped

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+#define TAG "network_protocol_redis_command_mget"
+
+struct mget_command_context {
+    char error_message[200];
+    bool has_error;
+    char *key;
+    size_t key_length;
+    size_t key_offset;
+    char **keys;
+    int entry_counter;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: As the `entries` field can be dropped from the struct, I would rename this in `keys_count`

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+typedef struct mget_command_context mget_command_context_t;
+
+NETWORK_PROTOCOL_REDIS_COMMAND_FUNCPTR_COMMAND_BEGIN(mget) {
+    protocol_context->command_context = slab_allocator_mem_alloc(sizeof(mget_command_context_t));
+
+    mget_command_context_t *mget_command_context = (mget_command_context_t*)protocol_context->command_context;
+    mget_command_context->has_error = false;
+    mget_command_context->keys = NULL;
+    mget_command_context->entries = NULL;
+    mget_command_context->entry_counter = 0;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: `reader_context->arguments.count - 1` can be assigned directly to the `keys_count` (after renaming) field and use it to allocate the memory below

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+    // there is no way to be sure that there is no corruption or data loss, so it's better to dramatically abort
+    // (in debug mode)
+    assert(mget_command_context->key_offset + chunk_length <= mget_command_context->key_length);
+
+    mget_command_context->key = mget_command_context->keys[argument_index] =
+            slab_allocator_mem_alloc(mget_command_context->key_length);
+
+    memcpy(mget_command_context->key, chunk_data, chunk_length);
+    mget_command_context->key_offset += chunk_length;
+    mget_command_context->keys[argument_index] = mget_command_context->key;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: This should be done in STREAM_END, STREAM_DATA can be invoked multiple times and therefore it would assign exactly the same value multiple times

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+            mget_command_context->entry_counter);
+
+    if (send_buffer_start == NULL) {
+        LOG_E(TAG, "buffer length incorrectly calculated, not enough space!");
+        goto end;
+    }
+
+    for(int i = 0; i < mget_command_context->entry_counter; i++) {
+        storage_db_entry_index_t *entry_index = mget_command_context->entries[i];
+
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c:199]: Add the code from STREAM_END that fetches the storage entry index here, just one very important catch is that the `entry_index` variable should be declared global in the scope of the function and then kept NULL if the `storage_db_entry_index_status_decrease_readers_counter` function has been invoked.

In this way at the `end` jump label, if `entry_index` is not NULL the `storage_db_entry_index_status_decrease_readers_counter` can be invoked to ensure that the readers counter of the entry index is always decremented as needed.

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+            send_buffer_end - send_buffer_start,
+            mget_command_context->entry_counter);
+
+    if (send_buffer_start == NULL) {
+        LOG_E(TAG, "buffer length incorrectly calculated, not enough space!");
+        goto end;
+    }
+
+    for(int i = 0; i < mget_command_context->entry_counter; i++) {
+        storage_db_entry_index_t *entry_index = mget_command_context->entries[i];
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: At this point it's necessary to have the key so this should become
```
char *key = mget_command_context->keys[i]
```

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+            send_buffer_start,
+            send_buffer_end - send_buffer_start,
+            mget_command_context->entry_counter);
+
+    if (send_buffer_start == NULL) {
+        LOG_E(TAG, "buffer length incorrectly calculated, not enough space!");
+        goto end;
+    }
+
+    for(int i = 0; i < mget_command_context->entry_counter; i++) {
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: It's probably better to rename the `i` variable in `key_index` or `index`

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+                error_send_buffer,
+                error_send_buffer_start - error_send_buffer) == NETWORK_OP_RESULT_OK;
+
+        goto end;
+    }
+
+
+//    char *send_buffer = NULL;
+//    char *send_buffer, *send_buffer_start, *send_buffer_end;
+    char send_buffer[512], *send_buffer_start, *send_buffer_end;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: To simplify the logic this `send_buffer` can be used only for the protocol elements, 64 bytes are probably plenty

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+    if (send_buffer_start == NULL) {
+        LOG_E(TAG, "buffer length incorrectly calculated, not enough space!");
+        goto end;
+    }
+
+    for(int i = 0; i < mget_command_context->entry_counter; i++) {
+        storage_db_entry_index_t *entry_index = mget_command_context->entries[i];
+
+        if (likely(entry_index)) {
+                send_buffer_length = min(entry_index->value_length + 32, STORAGE_DB_CHUNK_MAX_SIZE);
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: The send_buffer_length can't be changed, it indicates the length of the buffer defined at line 197

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+//                send_buffer_start = send_buffer = slab_allocator_mem_alloc(send_buffer_length);
+                send_buffer_end = send_buffer_start + send_buffer_length;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Can't be changed, it's the end of the buffer defined at line 197

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+
+                send_buffer_start = protocol_redis_writer_write_argument_blob_start(
+                        send_buffer_start,
+                        send_buffer_length,
+                        false,
+                        (int)entry_index->value_length);
+
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c:222]: It's necessary to do a network_send to send out the protocol data
```
    if (network_send(
            channel,
            send_buffer,
            send_buffer_start - send_buffer) != NETWORK_OP_RESULT_OK) {
        goto end;
    }
    send_buffer_start = send_buffer;
```

The code above will send out the data from send_buffer and then reset send_buffer_start to the very beginning of the buffer.

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+                // Build the chunks for the value
+                for(storage_db_chunk_index_t chunk_index = 0; chunk_index < entry_index->value_chunks_count; chunk_index++) {
+                    storage_db_chunk_info_t *chunk_info = storage_db_entry_value_chunk_get(entry_index, chunk_index);
+                    res = storage_db_entry_chunk_read(
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c:237]: It's better to allocate here the buffer needed to send out the data, there is a "fast mode" that can be used to bypass entirely the buffer allocation, if you take a look to the `network_protocol_redis_command_get.c` between line 201-311 but can be implemented at later stage as it's a small optimization.

e.g.
```
                    char *chunk_send_buffer = slab_allocator_mem_alloc(chunk_info->length);
                    res = storage_db_entry_chunk_read(
                             db,
                            chunk_info,
                            chunk_send_buffer);

                    if (!res) {
                        LOG_E(
                                TAG,
                                "[REDIS][MGET] Critical error, unable to read chunk <%u> long <%u> bytes",
                                chunk_index,
                                chunk_info->chunk_length);

                        goto end;
                    }

                    if (network_send(
                            channel,
                            chunk_send_buffer,
                            chunk_info->length) != NETWORK_OP_RESULT_OK) {
                        goto end;
                    }

                    slab_allocator_mem_free(chunk_buffer);
````

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+
+NETWORK_PROTOCOL_REDIS_COMMAND_FUNCPTR_COMMAND_FREE(mget) {
+    if (!protocol_context->command_context) {
+        return true;
+    }
+
+    mget_command_context_t *mget_command_context = (mget_command_context_t*)protocol_context->command_context;
+
+    slab_allocator_mem_free(mget_command_context->key);
+    slab_allocator_mem_free(mget_command_context->entries);
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Drop this

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+    slab_allocator_mem_free(mget_command_context->keys);
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Need to loop over the array of keys to free up the allocated memory

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+NETWORK_PROTOCOL_REDIS_COMMAND_FUNCPTR_COMMAND_FREE(mget) {
+    if (!protocol_context->command_context) {
+        return true;
+    }
+
+    mget_command_context_t *mget_command_context = (mget_command_context_t*)protocol_context->command_context;
+
+    slab_allocator_mem_free(mget_command_context->key);
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Should check if the `key` is `NULL` before freeing it up as the command might have been aborted before receiving the first argument or all the arguments have been received and all the keys are referenced in `keys` which need to be properly disposed (see comment below)

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+        storage_db_entry_index_status_increase_readers_counter(
+                entry_index,
+                &old_status);
+
+        if (unlikely(old_status.deleted)) {
+            entry_index = NULL;
+        }
+    }
+
+    mget_command_context->entries[argument_index] = entry_index;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: assign the `key` to the `keys` array and set `key` to NULL

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+    mget_command_context->entry_counter++;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Not needed, the count is assigned in COMMAND_BEGIN

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+                        LOG_E(
+                                TAG,
+                                "[REDIS][MGET] Critical error, unable to read chunk <%u> long <%u> bytes",
+                                chunk_index,
+                                chunk_info->chunk_length);
+
+                        goto end;
+                    }
+
+                    send_buffer_start += chunk_info->chunk_length;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Not needed

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 23]
```diff
@@ -0,0 +1,297 @@
...
+                }
+
+                // At this stage the entry index is not accessed further therefore the readers counter can be decreased. The
+                // entry_index has to be set to null to avoid that it's freed again at the end of the function
+                storage_db_entry_index_status_decrease_readers_counter(entry_index, NULL);
+                mget_command_context->entries[i] = entry_index = NULL;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Set only `entry_index` to NULL

**Review by Valkyrie00 — COMMENTED** [review]
Needs new review

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+    for(int key_index = 0; key_index < mget_command_context->keys_count; key_index++) {
+        storage_db_entry_index_status_t old_status;
+
+//        LOG_I(TAG, "%s %lu ",mget_command_context->keys[key_index],
+//              strlen(mget_command_context->keys[key_index]));
+
+        entry_index = storage_db_get_entry_index(
+                db,
+                mget_command_context->keys[key_index],
+                strlen(mget_command_context->keys[key_index]));
```
Comment (Valkyrie00) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c:194]: IDK how can i retrieve the right length of key

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 25]
```diff
@@ -0,0 +1,311 @@
...
+
+        if (likely(entry_index)) {
+                // Try to acquire a reader lock until it's successful or the entry index has been marked as deleted
+                storage_db_entry_index_status_increase_readers_counter(
+                        entry_index,
+                        &old_status);
+
+                // TODO: ??
+                if (unlikely(old_status.deleted)) {
```
Comment (Valkyrie00) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: A little review here pls 🙏 idk if i did it right

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 26]
```diff
@@ -0,0 +1,316 @@
...
+    if (mget_command_context->has_error) {
+        return true;
+    }
+
+    // The scenario tested in this assert can't happen, but if it does happen it's a bug in the protocol parser and
+    // there is no way to be sure that there is no corruption or data loss, so it's better to dramatically abort
+    // (in debug mode)
+    assert(mget_command_context->key_data->key_offset + chunk_length <= mget_command_context->key_data->key_length);
+
+    memcpy(mget_command_context->key_data->key, chunk_data, chunk_length);
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: When copying the data onto the key it's necessary to take into account key_offset because a key can be sent byte per byte, 10 bytes at time, etc.

It should be
```c
    memcpy(mget_command_context->key_data->key + mget_command_context->key_data->key_offset, chunk_data, chunk_length);
```

(better to split the line on multiple ones per argument as it's going to be fairly long after the change)

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c (hunk 25 above)
Comment (danielealbano) [reply src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: What's happend behind the scenes is that `storage_db_entry_index_status_increase_readers_counter` will internally increase a counter to `prevent` the entry from being deleted (by a SET that replace the key or a DELETE) and ensure that it hasn't been deleted while the counter was being incremented.

Normally `storage_db_get_entry_index` will not return the entry index if deleted but it might get deleted by another thread after it's returned, so there is this extra check.

The `entry_index` returned by `storage_db_get_entry_index` is guaranteed to exist in memory until a fiber switch if it;s not locked by `storage_db_entry_index_status_increase_readers_counter` but it's not guaranteed to do not have been marked deleted by another thread (which means it's going to be deleted very soon if there are readers or within the next fiber switch if there aren't).

So long story short: once the entry index is fetched and the readers counter is increased it's necessary to check if the status fetched from the call is reporting the entry to be deleted.

If it's reported as deleted it's not necessary to decrease the readers counter, this is done automatically internally by `storage_db_entry_index_status_increase_readers_counter`.

If it's deleted it's better to set the entry_index to `NULL`.

The code should always terminate the if and have another if that checks again if `entry_index` is `NULL` so if it set to `NULL` by the previous process the code will flow normally as there would be no mach in the storage db at all.

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 27]
```diff
@@ -0,0 +1,316 @@
...
+                storage_db_entry_index_status_increase_readers_counter(
+                        entry_index,
+                        &old_status);
+
+                // TODO: ??
+                if (unlikely(old_status.deleted)) {
+                    entry_index = NULL;
+                }
+
+                send_buffer_start = protocol_redis_writer_write_argument_blob_start(
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c:217]: because of the very long explanation above, the previous
```
if (likely(entry_index)) {
```

Should be terminated before this line and another exactly the same should be added there so if the entry_index has been deleted meanwhile it was being locked the code will flow as if there was no key available at all.

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 28]
```diff
@@ -0,0 +1,316 @@
...
+                                chunk_info->chunk_length);
+
+                        goto end;
+                    }
+
+                    if (network_send(
+                            channel,
+                            chunk_send_buffer,
+                            chunk_info->chunk_length) != NETWORK_OP_RESULT_OK) {
+                        goto end;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c:257]: Better to add a `slab_allocator_mem_free(chunk_send_buffer);` before the `goto end;` to avoid a memory leak

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 29]
```diff
@@ -0,0 +1,316 @@
...
+                            chunk_send_buffer);
+
+                    if (!res) {
+                        LOG_E(
+                                TAG,
+                                "[REDIS][MGET] Critical error, unable to read chunk <%u> long <%u> bytes",
+                                chunk_index,
+                                chunk_info->chunk_length);
+
+                        goto end;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c:249]: Better to add a `slab_allocator_mem_free(chunk_send_buffer);` before the `goto end;` to avoid a memory leak

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 30]
```diff
@@ -0,0 +1,316 @@
...
+    }
+
+    mget_command_context_t *mget_command_context = (mget_command_context_t*)protocol_context->command_context;
+
+    if (mget_command_context->key_data == NULL) {
+        slab_allocator_mem_free(mget_command_context->key_data);
+    }
+
+    for(int key_index = 0; key_index < mget_command_context->keys_count; key_index++) {
+        slab_allocator_mem_free(mget_command_context->keys[key_index]);
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Potentially a key might be null because the mget command aborted before receiving all the keys, it's better to free the key only if the value is not null (I dropped a related command at the line where the memory is allocated)

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 31]
```diff
@@ -0,0 +1,316 @@
...
+#include "network/network.h"
+#include "worker/worker_stats.h"
+#include "worker/worker_context.h"
+
+#define TAG "network_protocol_redis_command_mget"
+
+struct mget_command_key_data {
+    char *key;
+    size_t key_length;
+    size_t key_offset;
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: A tiny tiny and optional improvement would be to move `key_offset` under `mget_command_context` because it's need only while receiving the data and only one at time, we don't need to preserve it at all.

Or we can potentially even use just `key_length` .

The reason is that if someone requests 100000 keys the memory needed for the struct above will be 24 bytes so 2.4MB will be required, if we drop the key_offset from there moving it to `mget_command_context` (or use key_length in its place) it will only need 16 bytes and therefore 1.6MB.

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 32]
```diff
@@ -0,0 +1,316 @@
...
+};
+typedef struct mget_command_context mget_command_context_t;
+
+NETWORK_PROTOCOL_REDIS_COMMAND_FUNCPTR_COMMAND_BEGIN(mget) {
+    protocol_context->command_context = slab_allocator_mem_alloc(sizeof(mget_command_context_t));
+
+    mget_command_context_t *mget_command_context = (mget_command_context_t*)protocol_context->command_context;
+    mget_command_context->has_error = false;
+    mget_command_context->keys_count = reader_context->arguments.count - 1;
+    mget_command_context->keys = slab_allocator_mem_alloc(mget_command_context->keys_count * sizeof(mget_command_key_data_t*));
```
Comment (danielealbano) [inline src/network/protocol/redis/command/network_protocol_redis_command_mget.c]: Better to zero this memory upon allocation, it makes easy to identify that there is memory to be freed in NETWORK_PROTOCOL_REDIS_COMMAND_FUNCPTR_COMMAND_FREE (there is a related comment there).

**Review by danielealbano — CHANGES_REQUESTED** [review]
Spotted a few places where it's necessary to fix things and suggested an optional tiny improvement (not sure if it's worth because better improvements can be done but they are for another PR :))

File: src/network/protocol/redis/command/network_protocol_redis_command_mget.c [hunk 33]
```diff
@@ -0,0 +1,316 @@
...
+        return true;
+    }
+
+    // The scenario tested in this assert can't happen, but if it does happen it's a bug in the protocol parser and
+    // there is no way to be sure that there is no corruption or data loss, so it's better to dramatically abort
+    // (in debug mode)
+    assert(mget_command_context->key_data->key_offset + chunk_length <= mget_command_context->key_data->key_length);
+
+    memcpy(mget_command_context->key_data->key, chunk_data, chunk_length);
+    mget_command_context->key_data->key_offset += chunk_length;
```
[the remark this answers is not in the document]
Comment (danielealbano) [reply src/network/protocol/redis/command/network_protocol_redis_command_mget.c:125]: key_length is actually set at line 104 so no need (thanks @Valkyrie00 for catching that my suggestion was incorrect!)
